src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,341 @@
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		"Script - ANY",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleNofK(read_view_t* view)
+{
+	// parse data
+	ctx->nativeScript.nOfK.n = parse_u4be(view);
+	ctx->remainingScripts[ctx->level] = parse_u4be(view);
+	ctx->type[ctx->level] = NATIVE_COMPLEX_SCRIPT_N_OF_K;
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("level = %u, required scripts = %u, remaining scripts = %u", ctx->level, ctx->nativeScript.nOfK.n, ctx->remainingScripts[ctx->level]);
+	
+	// validate that the received requiredScripts count makes sense
+	VALIDATE(ctx->remainingScripts[ctx->level] >= ctx->nativeScript.nOfK.n, ERR_INVALID_DATA);
+
+	// max possible length 32: "Confirm x of x"
+	// where x is 2^32-1
+	char text[33];
+	explicit_bzero(text, SIZEOF(text));
+	snprintf(text, SIZEOF(text), "Confirm %d of %d", ctx->nativeScript.nOfK.n, ctx->remainingScripts[ctx->level]);
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		"Script - N of K",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	VALIDATE(expectsRequest(), ERR_INVALID_STATE);
+
+	// check if we can increase the level without breaking the MAX_SCRIPT_DEPTH constraint
+	VALIDATE(ctx->level + 1 < MAX_SCRIPT_DEPTH, ERR_DATA_TOO_LARGE);
+	ctx->level++;
+
+	uint8_t nativeScriptType = parse_u1be(view);
+	TRACE("native complex script type = %u", nativeScriptType);
+
+	switch(nativeScriptType) {
+#	define  CASE(TYPE, HANDLER) case TYPE: HANDLER(view); break;
+		CASE(NATIVE_SCRIPT_ALL, deriveNativeScriptHash_handleAll);
+		CASE(NATIVE_SCRIPT_ANY, deriveNativeScriptHash_handleAny);
+		CASE(NATIVE_SCRIPT_N_OF_K, deriveNativeScriptHash_handleNofK);
+		default: THROW(ERR_INVALID_DATA);
+#	undef   CASE
+	}
+}
+
+// Simple native scripts
+
+void deriveNativeScriptHash_handleDeviceOwnedPubkey(read_view_t* view)
+{
+	bip44_path_t *path = &ctx->nativeScript.pubkeyDeviceOwned.path;
+	view_skipBytes(view, bip44_parseFromWire(path, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(view)));
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("Pubkey given by path:");
+	BIP44_PRINTF(path);
+	PRINTF("\n");
+
+	ui_displayPathScreen(
+		"Script - key path",
+		&ctx->nativeScript.pubkeyDeviceOwned.path,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleThirdPartyPubkey(read_view_t* view)
+{
+	uint8_t *hash = ctx->nativeScript.pubkeyThirdParty.hash;
+	VALIDATE(view_remainingSize(view) == ADDRESS_KEY_HASH_LENGTH, ERR_INVALID_DATA);
+	view_memmove(hash, view, ADDRESS_KEY_HASH_LENGTH);
+
+#	ifdef DEVEL
+	TRACE("Pubkey given by hash:");
+	for (int i = 0; i < ADDRESS_KEY_HASH_LENGTH; i++) {
+		PRINTF("%x", hash[i]);
+	}
+	PRINTF("\n");
+#	endif
+
+	ui_displayHexBufferScreen(
```

Comment:
Don't we use bech32 encoding for displaying keys?

Reply:
I switched to bech32, but wasn't sure what prefix to use, tried to look for other usecases in the codebase, but that didn't help. For now went with the `pubkey` prefix, but I'll welcome any suggestions.

Reply:
This should be dealt with across the whole codebase. A very reasonable task for you, btw.; consult with Gabo or on the small daily standup the individual instances where we are displaying keys. The bech32 prefixes are here:
https://github.com/cardano-foundation/CIPs/blob/master/CIP-0005/CIP-0005.md

Reply:
@janmazak so I assume by [this message in Slack from Gabo](https://vacuumlabs.slack.com/archives/CETDV69MF/p1628162445019000?thread_ts=1627911175.004900&cid=CETDV69MF) that I should check/make this change across the Ledger codebase, right?

Reply:
Yes.

Reply:
Created an issue for this: https://github.com/vacuumlabs/ledger-app-cardano-shelley/issues/90